pkg/controller/reconciler/deploymentresource_reconciler.go

```diff
@@ -0,0 +1,374 @@
+			err = r.Client.Status().Update(ctx, deploymentResource)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{Requeue: true, RequeueAfter: r.requeueDelay()}, nil
+		}
+
+		// If we get here, the operation was a success. Update the status and continue.
+		deploymentResource.Status.Operation = nil
+		deploymentResource.Status.Phrase = radappiov1alpha3.DeploymentResourcePhraseDeleted
+		err = r.Client.Status().Update(ctx, deploymentResource)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		logger.Info("Resource is deleted.")
+
+		// At this point we've cleaned up everything. We can remove the finalizer which will allow deletion of the
+		// DeploymentResource
+		if controllerutil.RemoveFinalizer(deploymentResource, DeploymentResourceFinalizer) {
+			deploymentResource.Status.ObservedGeneration = deploymentResource.Generation
+			deploymentResource.Status.Phrase = radappiov1alpha3.DeploymentResourcePhraseDeleted
+			err = r.Client.Update(ctx, deploymentResource)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// If we get here, this was an unknown operation kind. This is a bug in our code, or someone
+	// tampered with the status of the object. Just reset the state and move on.
+	logger.Error(fmt.Errorf("unknown operation kind: %s", deploymentResource.Status.Operation.OperationKind), "Unknown operation kind.")
+
+	deploymentResource.Status.Operation = nil
+	deploymentResource.Status.Phrase = radappiov1alpha3.DeploymentResourcePhraseFailed
+	err := r.Client.Status().Update(ctx, deploymentResource)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *DeploymentResourceReconciler) reconcileDelete(ctx context.Context, deploymentResource *radappiov1alpha3.DeploymentResource) (ctrl.Result, error) {
+	logger := ucplog.FromContextOrDiscard(ctx)
+
+	logger.Info("Resource is being deleted.", "resourceId", deploymentResource.Spec.Id)
+
+	// Since we're going to reconcile, update the observed generation.
+	//
+	// We don't want to do this if we're in the middle of an operation, because we haven't
+	// fully processed any status changes until the async operation completes.
+	deploymentResource.Status.ObservedGeneration = deploymentResource.Generation
+
+	// NOTE: The following is a workaround for Radius API behavior. Since deleting
+	// an application or environment can leave hanging resources, we need to make sure to
+	// delete these resources before deleting the application or environment.
+
+	// Check other resources that depend on this resource.
+	// List all DeploymentResource objects in the same namespace
+	// that have the same rootFileName.
+	deploymentResourceList := &radappiov1alpha3.DeploymentResourceList{}
+	err := r.Client.List(ctx, deploymentResourceList, client.InNamespace(deploymentResource.Namespace), client.MatchingFields{rootFileNameField: deploymentResource.Spec.RootFileName})
```

Comment:
Is this how we're associating a DT with its DRs? 

We should be using owner references for this. https://book.kubebuilder.io/reference/watching-resources/secondary-owned-resources

Reply:
Or is this a workaround for ordering constraints?

Reply:
If that's the case we should still user owner reference, but without the automatic cascading delete: https://book.kubebuilder.io/reference/watching-resources/secondary-owned-resources